File: mid/Array/33. Search in Rotated Sorted Array.py
# https://leetcode.com/problems/search-in-rotated-sorted-array/solutions/3879263/100-binary-search-easy-video-o-log-n-optimal-solution/
# Time Complexity : O(log⁡n)
# Space Complexity : O(1)
# Binary Search

class Solution():
    def search(self, nums, target):
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid

            if nums[mid] >= nums[left]: # left rotated => nums= 4 5 6 7 0 1 2 (mid=7, target=0)
                # in ascending order side
                if nums[left] <= target and target < nums[mid]: # mid=7, target=5, get rid of the right side.
                    right = mid - 1 
                else:
                    left = mid + 1  # mid=7, target=1, get rid of the left side.
                    
            else: # right rotated => nums= 6 7 0 1 2 3 4   (mid=1, target=0)
                # in ascending order side
                if nums[mid] < target and target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        # cannot find the target value
        return -1
    
    
if __name__ == '__main__':
    nums = [4,5,6,7,0,1,2]
    target = 0
    function = Solution()
    ans = function.search(nums, target)
    print('ans= ', ans)

File: mid/Array/33. Search in Rotated Sorted Array.cpp
#include<array> 
#include<iostream>    
#include <vector>

// https://leetcode.com/problems/search-in-rotated-sorted-array/solutions/3879263/100-binary-search-easy-video-o-log-n-optimal-solution/
// Time Complexity : O(log⁡n)
// Space Complexity : O(1)
// Binary Search

class Solution {
public:
    int search(std::vector<int>& nums, int target) {
        int low = 0, high = nums.size() - 1;

        while (low <= high) {
            int mid = (low + high) / 2;

            if (nums[mid] == target) { // 剛好中間點就是答案就
                return mid;
            }
            // 左半邊是有序的 (left portion is non-rotated.)
            if (nums[low] <= nums[mid]) { 
                // 如果 target 在左半邊的範圍內，則繼續在左半邊進行搜尋 (search left portion (non-rotated))
                if (target >= nums[low] && target < nums[mid]) {
                    high = mid - 1; // 縮小右邊
                } else { // 否則，轉向右半邊進行搜尋
                    low = mid + 1; // 縮小左邊
                }
            } else { // 否則，則右半邊是有序的 (right portion is non-rotated/ left portion is rotated).
                // 如果 target 在右半邊的範圍內，則繼續在右半邊進行搜尋
                if (target > nums[mid] && target <= nums[high]) {
                    low = mid + 1; 
                } else { // 否則，轉向左半邊進行搜尋
                    high = mid - 1; 
                }
            }
        }
        return -1;
    }
};



int main() {
    Solution solution;

    // Example 1
    std::vector<int> nums1 = {4, 5, 6, 7, 0, 1, 2};
    int target1 = 0;
    int result1 = solution.search(nums1, target1);
    std::cout << "Example 1: Result = " << result1 << std::endl;

    // Example 2
    std::vector<int> nums2 = {4, 5, 6, 7, 0, 1, 2};
    int target2 = 3;
    int result2 = solution.search(nums2, target2);
    std::cout << "Example 2: Result = " << result2 << std::endl;

    // Add more test cases as needed

    return 0;
}



